core/src/Kokkos_Macros.hpp

```diff
@@ -633,7 +633,8 @@ static constexpr bool kokkos_omp_on_host() { return false; }
 #define KOKKOS_ENABLE_CUDA_LDG_INTRINSIC
 #endif
 
-#if defined(KOKKOS_ENABLE_CXX17) || defined(KOKKOS_ENABLE_CXX20)
+#if (defined(KOKKOS_ENABLE_CXX17) || defined(KOKKOS_ENABLE_CXX20)) && \
+    (__has_cpp_attribute(nodiscard) >= 201907)
```

Comment:
I think we should just have
```
#if defined(__has_cpp_attribute) && __has_cpp_attribute(nodiscard) >= 201907
KOKKOS_ATTRIBUTE_NODISCARD
#endif
```
in front of the constructors and leave the macro alone otherwise.

Reply:
Updated according to this suggestion.